Keep explicit ticklabels in sync with ticks from FixedLocator

efiring opened the pull request:
## PR Summary
Using FixedFormatter with ticks from FixedLocator causes mislabeled ticks if the nbins kwarg of the locator triggers subsampling.  This PR addresses that in the common case where the `Axis.set_ticklabels` method is used, by setting a FuncFormatter instead of a FixedFormatter.

Closes #11937.
## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

QuLogic commented:
xref #15005

anntzer commented on lib/matplotlib/axis.py:
```diff
@@ -1626,14 +1631,24 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
+            self.set_major_formatter(formatter)
             ticks = self.get_major_ticks()
+
+        self._update_ticks()
```
do you really need to call update_ticks() here?  and if you do, shouldn't this go *before* get_major_ticks()/get_minor_ticks()?  (otherwise you'd be possibly invalidating `ticks` even before iterating on them...)

anntzer commented on lib/matplotlib/ticker.py:
```diff
@@ -379,18 +379,28 @@ class FuncFormatter(Formatter):
     position ``pos``), and return a string containing the corresponding
-    tick label.
+    tick label.  The function may take additional arguments that are
+    supplied upon instantiation.
```
"appended" (they go after `x`, `pos`, not before).

anntzer commented on lib/matplotlib/axis.py:
```diff
@@ -1626,14 +1631,24 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
+                   else self.get_major_locator())
+        if isinstance(locator, mticker.FixedLocator):
+            tickd = {loc: lab for loc, lab in zip(locator.locs, ticklabels)}
+            formatter = mticker.FuncFormatter(self._format_with_dict, tickd)
```
I'm not sure whether we want or not to make this extension to the FuncFormatter API (perhaps?) but if not, I think you can get away with
```
@staticmethod
def _format_with_ticks(tickd, x, pos): ...  # move tickd to first

...

formatter = FuncFormatter(functools.partial(self._format_with_ticks, tickd))
```
as the `partial` object will stay picklable (I think this is more or less the canonical way to bind some arguments while keeping picklability).

efiring commented:
@anntzer I think I have addressed all your points.  I had tried using partial earlier, but hadn't gotten it right.  _update_ticks is overkill, but some of its function is essential to passing the pre-existing `test_axes.py::test_set_get_ticklabels`, so I put the needed parts directly into `set_ticklabels`.

anntzer commented on lib/matplotlib/axis.py:
```diff
@@ -1626,14 +1632,28 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
-        for tick_label, tick in zip(ticklabels, ticks):
+        for pos, (loc, tick) in enumerate(zip(locs, ticks)):
+            tick.update_position(loc)
+            tick_label = formatter(tick._loc, pos)
```
is `tick._loc` just `loc`? or not?

efiring replied:
At this point they should be identical; but yes, it might look better and be slightly more efficient with `loc`.

QuLogic commented on lib/matplotlib/tests/test_axes.py:
```diff
@@ -4452,6 +4452,18 @@ def test_set_get_ticklabels():
+    # test issue 11937
+    fig, ax = plt.subplots()
+    ax.plot(np.arange(10))
+    ax.xaxis.set_ticks(np.arange(10) + 0.1)
+    ax.locator_params(nbins=5)
+    ax.xaxis.set_ticklabels([c for c in "bcdefghijk"])
```
Putting these together would make it a bit clearer:
```suggestion
    ax.xaxis.set(ticks=np.arange(10) + 0.1)
                 ticklabels=[c for c in "bcdefghijk"])
    ax.locator_params(nbins=5)
```

efiring replied:
The use of ``set`` here would work only by accident, because the kwargs are sorted in inverse order, so that set_ticks will be called before set_ticklabels.  I will stick with the individual function calls.

QuLogic replied:
This is the closest analogue to `plt.xticks(locs, labels)`, so if you're saying it might be broken, then I think that's not good. Especially given the recent tweet about `.set()`...

efiring replied:
I don't use twitter and haven't seen the tweet.

I'm not saying the behavior is currently broken, just that it works in this case (I think--I haven't tried it) because the names are sorted in inverse order, and the "s" in "ticks" makes it sort before the first "l" in "ticklabels".  It's an accident.  It might be better to make it explicit by adding a "ticks" entry to Artist._prop_order to boost its priority.  The _prop_order dict provides a way of setting priority at a higher level than the alphanumeric sort.   Presently, it serves only to give 'color' low priority, so it is executed last.

QuLogic replied:
[The is the tweet](https://twitter.com/matplotlib/status/1256364300492513281) that @story645 mentioned.

For users, working by accident and working by design are the same thing. So I meant if _this PR_ breaks it, that would not be good. However, it doesn't, so we're fine, even if it's working by accident right now.

anntzer replied:
I guess I'll take advantage of this to point to https://github.com/matplotlib/matplotlib/pull/16328 :)

jklymak commented:
As discussed on the call, it would be good if there was a lock on `set_ticklabels` that said it had to be the same length as `set_ticks`.

jklymak commented on lib/matplotlib/axis.py:
```diff
@@ -1635,6 +1635,11 @@ def set_ticklabels(self, ticklabels, *, minor=False, **kwargs):
         locator = (self.get_minor_locator() if minor
                    else self.get_major_locator())
         if isinstance(locator, mticker.FixedLocator):
+            if len(locator.locs) != len(ticklabels):
```
L 1620 probably needs something like "Number of labels must be the same as number of ticks set by `set_ticks` or the `FixedLocator`"?

efiring replied:
I added a reference to set_ticks.

jklymak replied:
That helps, but I mean above in the Parameter description.  Its a bit unclear as to what is meant by "tick labels".   maybe:

```
ticklabels : sequence of str or of `.Text`\s
            List of texts for tick labels, one for each tick defined in `.axis.set_ticks`.
```

efiring commented:
Adding the check for a match turned up a blatant error in our image_annotated_heatmap example.  I wonder how many such mismatches are out in the wild, and not as obvious as in our example?  We'll see...

efiring commented:
The appveyor failure appears to be caused by the crash of the VM running the test.

efiring commented:
@jklymak I have augmented the ticklabels kwarg description in the docstring, and updated a statement in the API docs about `set_ticks` and `set_ticklabels`.

jklymak commented:
I'll leave here since there is a mailing list discussion, but anyone may merge after that is done.
